payroll/models/hr_payslip.py

```diff
@@ -575,44 +570,54 @@ def _compute_payslip_line(self, rule, localdict, lines_dict):
-                payslip.employee_id.id,
-                payslip.get_current_contract_dict(contract, contracts),
-                self.env,
-            )
-            # set up localdict with current contract and employee values
-            localdict = dict(
-                baselocaldict, employee=contract.employee_id, contract=contract
-            )
-            for rule in self._get_salary_rules(contracts, payslip):
-                localdict["result"] = None
-                localdict["result_qty"] = 1.0
-                localdict["result_rate"] = 100
-                localdict["result_name"] = None
-                # check if the rule can be applied
-                if rule._satisfy_condition(localdict) and rule.id not in blacklist:
-                    localdict, lines_dict = payslip._compute_payslip_line(
-                        rule, localdict, lines_dict
-                    )
-                else:
-                    # blacklist this rule and its children
-                    blacklist += [id for id, seq in rule._recursive_search_of_rules()]
-            # call localdict_hook
-            localdict = self.localdict_hook(localdict, payslip)
-            # reset "current_contract" dict
-            baselocaldict["current_contract"] = {}
+        for payslip in self:
+            contracts = payslip._get_employee_contracts()
+            baselocaldict = payslip._get_baselocaldict(contracts)
+            for contract in contracts:
+                # assign "current_contract" dict
+                baselocaldict["current_contract"] = BrowsableObject(
+                    payslip.employee_id.id,
+                    payslip.get_current_contract_dict(contract, contracts),
+                    payslip.env,
+                )
+                # set up localdict with current contract and employee values
+                localdict = dict(
+                    baselocaldict, employee=contract.employee_id, contract=contract
+                )
+                for rule in payslip._get_salary_rules():
+                    localdict["result"] = None
+                    localdict["result_qty"] = 1.0
+                    localdict["result_rate"] = 100
+                    localdict["result_name"] = None
+                    # check if the rule can be applied
+                    if rule._satisfy_condition(localdict) and rule.id not in blacklist:
+                        localdict, _dict = payslip._compute_payslip_line(
+                            rule, localdict, lines_dict
+                        )
+                    else:
+                        # blacklist this rule and its children
+                        blacklist += [
+                            id for id, seq in rule._recursive_search_of_rules()
+                        ]
+                    lines_dict.update(_dict)
+                    # call localdict_hook
+                    localdict = payslip.localdict_hook(localdict)
```

Comment:
Please note that localdict_hook was outside the loop in the previous implementation. 

For my implementations I need it outside the loop. But you have me the idea to have two hooks, one inside the loop and one outside it. 

What do you think? They could be named: after_localdict_hook and onloop_localdict_hook. I'm open to better names haha.

Reply:
Ahh yes, sorry. This is a mistake on my part. As for having both the inner and outer hook I'm don't know how much more useful it is than just inheriting _compute_payslip_line().